fix: change favicon per theme

DiogoSoaress opened the pull request:
## What it solves

Resolves https://github.com/safe-global/web-core/issues/1123

## How this PR fixes it
Implements a hook to detect the system's preferred color scheme and subscribes to that media query changes

## How to test it
Go to your OS preferences and change the dark/light theme.
Observe the favicon in the browsers tab changing accordingly.

## Screenshots
![Screenshot 2022-12-12 at 15 41 49](https://user-images.githubusercontent.com/32431609/207073676-26224732-4361-4680-810b-4a37d11ba50e.png)

schmanu commented on src/hooks/__tests__/usePrefersColorScheme.test.ts:
```diff
@@ -0,0 +1,13 @@
+import { renderHook } from '@/tests/test-utils'
+
+describe('usePrefersColorScheme', () => {
+  it('should return the initial color scheme', () => {
```
Could we add more tests for the listeners of the media query which would set the result to dark / light?

DiogoSoaress replied:
Added. I found it a bit difficult to mock the media query event without adding the dependency. Maybe we can look at it quickly together

schmanu replied:
I can have a look later to do it as described here:
https://jestjs.io/docs/26.x/manual-mocks#mocking-methods-which-are-not-implemented-in-jsdom

iamacook commented on src/hooks/usePrefersColorScheme.ts:
```diff
@@ -0,0 +1,33 @@
+
+type ColorSchemeType = 'dark' | 'light' | 'no-preference'
+
+const usePrefersColorScheme = (): ColorSchemeType => {
```
We should also take the dark theme into account. We have a very similar hook called `useDarkMode`.

DiogoSoaress replied:
You mean considering the app's theme to affect the favicon?

iamacook replied:
Yes. The app theme may not match that of the OS.

DiogoSoaress replied:
Exactly for that reason, we should only make the favicon contrast with the OS theme. The app theme does not  affect the browser tabs color.

iamacook replied:
I see. Oversight on my behalf.

Does this work nicely with the pairing module/WC? I don't have an iPhone so I can't test it.

iamacook replied:
And the icons are clearly visible in light/dark mode on WC/app?

DiogoSoaress replied:
Yes. Those icons have the green circle background (I didn't change it)

iamacook commented on src/components/common/MetaTags/index.tsx:
```diff
@@ -41,8 +41,6 @@ const MetaTags = ({ prefetchUrl }: { prefetchUrl: string }) => (
     {/* Favicons */}
     <link rel="shortcut icon" href="/favicons/favicon.ico" />
     <link rel="apple-touch-icon" sizes="180x180" href="/favicons/apple-touch-icon.png" />
-    <link rel="icon" type="image/png" sizes="32x32" href="/favicons/favicon-32x32.png" />
-    <link rel="icon" type="image/png" sizes="16x16" href="/favicons/favicon-16x16.png" />
     <link rel="mask-icon" href="/favicons/safari-pinned-tab.svg" color="#000" />
```
Do we not need to update these other favicons?

DiogoSoaress replied:
I've added dark versions to the ones that were subject to poor contrast in browser's tabs.
If I get it correctly the ones I left in `<MetaTags>` are for other uses.

iamacook replied:
I'm still seeing the old icon in Firefox:
![image](https://user-images.githubusercontent.com/20442784/207297325-ccd331e0-58e4-40f4-b464-00c91a231091.png)

I would suggest we test adding an icon to the home screen of an iPhone, as well as pinning a tab in Safari.

iamacook replied:
I opened it in a Private Window though.

iamacook commented on src/hooks/usePrefersColorScheme.ts:
```diff
@@ -0,0 +1,33 @@
+    const isDark = window.matchMedia('(prefers-color-scheme: dark)')
+    const isLight = window.matchMedia('(prefers-color-scheme: light)')
+
+    const darkListener = ({ matches }: MediaQueryListEvent) => {
+      matches && setPreferredColorScheme('dark')
+    }
+    const lightListener = ({ matches }: MediaQueryListEvent) => {
+      matches && setPreferredColorScheme('light')
+    }
```
You could curry these to reduce repetition, e.g.

```suggestion
    const listener = (type: ColorSchemeType) = ({ matches }: MediaQueryListEvent) => {
      matches && setPreferredColorScheme(type)
    }
    
    const darkListener = listener('dark')
    const lightListener = listener('light')
```

DiogoSoaress replied:
Good suggestion! Done

iamacook commented on src/pages/_app.tsx:
```diff
@@ -77,11 +78,16 @@ interface WebCoreAppProps extends AppProps {
       <Head>
         <title key="default-title">Safe</title>
         <MetaTags prefetchUrl={GATEWAY_URL} />
+        <link rel="icon" type="image/png" sizes="32x32" href={`/favicons/favicon${darkPath}-32x32.png`} />
+        <link rel="icon" type="image/png" sizes="16x16" href={`/favicons/favicon${darkPath}-16x16.png`} />
```
Can we not keep these and associated logic inside `MetaTag`?

DiogoSoaress replied:
I tried to include the logic inside <MetaTag /> component but it was not taking effect. I tried to log from inside that component but unsuccessfully. Happy to look at it together if we want to move these `<link>` back there

katspaugh commented on src/hooks/usePrefersColorScheme.ts:
```diff
@@ -0,0 +1,35 @@
+  useEffect(() => {
+    if (typeof window.matchMedia !== 'function') return
+
+    const isDark = window.matchMedia('(prefers-color-scheme: dark)')
+    const isLight = window.matchMedia('(prefers-color-scheme: light)')
```
Aren't they mutually exclusive?

DiogoSoaress replied:
From my understanding of https://w3c.github.io/csswg-drafts/mediaqueries-5/#prefers-color-scheme they are not mutually exclusive but correct me if I'm wrong. The user can prefer `dark` or `light` or have no preference.

katspaugh replied:
Yeah, but you don't have a third icon for `!isDark && !isLight`, do you?
Also from the docs you linked (emphasis mine):
> `light`
> Indicates that user has expressed the preference for a page that has a light theme (dark text on light background), **or has not expressed an active preference (and thus should receive the "web default" of a light theme)**.

katspaugh commented on src/pages/_app.tsx:
```diff
@@ -77,11 +78,16 @@ interface WebCoreAppProps extends AppProps {
 
 const WebCoreApp = ({ Component, pageProps, emotionCache = clientSideEmotionCache }: WebCoreAppProps): ReactElement => {
+  const prefersColorScheme = usePrefersColorScheme()
+  const darkPath = prefersColorScheme === 'dark' ? '-dark' : ''
```
`darkPath` is not the best name if it can be either dark or light.

katspaugh reviewed (changes requested):
You propose adding 128 lines of JS and an external library to change a favicon.
From the maintainability standpoint, I would much prefer a low-tech solution like an icon with a background.

katspaugh commented:
There's a CSS-only solution too, have you tried it?
https://stackoverflow.com/a/60158619/352796

DiogoSoaress commented:
> You propose adding 128 lines of JS and an external library to change a favicon.

I know. Well, the changes are more numerous given the added hook and the test coverage. The dev dependency is also only used for tests and it was added following the code review.

I will talk with @liliiaorlenko to get a `svg` favicon to explore the CSS-only solution and to get her feedback on changing the favicon to a round background that would have good contrast regardless of the OS preferred theme.

DiogoSoaress commented:
> There's a CSS-only solution too, have you tried it?

I was trying this implementation but I don't think it suits our use case give it builds over the assumption that nowadays you can use an SVG as a favicon for your website, which is hardly true.
https://caniuse.com/link-icon-svg

I also got the favicon with the green background from the designer. I will create a different PR for it so we can compare the 2 solutions

DiogoSoaress commented:
I've addressed the comments and opened an alternative implementation to this PR using the favicon with the green background https://github.com/safe-global/web-core/pull/1376

katspaugh commented:
It wasn't necessary to spend time addressing the comments. Closing in favor of the other PR.
